### ISSUE TITLE
`Unsupported: call_method SizeVariable() numel [] {}` when compiling a function sampling from Categorical

### ISSUE THREAD

**Issue by KareemYousrii** [pytorch/pytorch#106407]
### 🐛 Describe the bug

Compiling the simple function below

```
import torch
from torch.distributions.categorical import Categorical
@torch.compile(fullgraph=True, mode='reduce-overhead')
def func():
    sample = Categorical(torch.rand(10, 5)).sample()
```

fails with `Unsupported: call_method SizeVariable() numel [] {}`

### Error logs

Traceback (most recent call last):
  File "/scratch/ahmedk/hard_ar_sl/ExprStarcraftShortestPath_hard/compile_sampling.py", line 8, in <module>
    func()
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 286, in _fn
    return fn(*args, **kwargs)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 439, in catch_errors
    return callback(frame, cache_size, hooks, frame_state)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 127, in _fn
    return fn(*args, **kwargs)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 360, in _convert_frame_assert
    return _compile(
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 430, in _compile
    out_code = transform_code_object(code, transform)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/bytecode_transformation.py", line 1000, in transform_code_object
    transformations(instructions, code_options)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 415, in transform
    tracer.run()
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2020, in run
    super().run()
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 704, in run
    and self.step()
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 664, in step
    getattr(self, inst.opname)(inst)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 386, in wrapper
    return inner_fn(self, inst)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 1095, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 555, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 333, in call_function
    return super().call_function(tx, args, kwargs)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 287, in call_function
    return super().call_function(tx, args, kwargs)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 120, in call_function
    return tx.inline_user_function_return(
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 591, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2125, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2222, in inline_call_
    tracer.run()
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 704, in run
    and self.step()
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 664, in step
    getattr(self, inst.opname)(inst)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 386, in wrapper
    return inner_fn(self, inst)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 1095, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 555, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/variables/misc.py", line 558, in call_function
    return self.obj.call_method(tx, self.name, args, kwargs).add_options(self)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/variables/lists.py", line 510, in call_method
    return super().call_method(tx, name, args, kwargs)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/variables/lists.py", line 422, in call_method
    return super().call_method(tx, name, args, kwargs)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/variables/lists.py", line 119, in call_method
    return super().call_method(tx, name, args, kwargs)
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/variables/base.py", line 253, in call_method
    raise unimplemented(f"call_method {self} {name} {args} {kwargs}")
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/_dynamo/exc.py", line 134, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: call_method SizeVariable() numel [] {}

from user code:
   File "/scratch/ahmedk/hard_ar_sl/ExprStarcraftShortestPath_hard/compile_sampling.py", line 6, in func
    sample = Categorical(torch.rand(10, 5)).sample()
  File "/space/ahmedk/anaconda3/envs/pt2/lib/python3.10/site-packages/torch/distributions/categorical.py", line 118, in sample
    samples_2d = torch.multinomial(probs_2d, sample_shape.numel(), True).T

You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230603
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.0 (default, Mar  3 2022, 09:58:08) [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-48-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA RTX A6000
GPU 1: NVIDIA GeForce RTX 2080 Ti
GPU 2: NVIDIA RTX A6000

Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   48 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          32
On-line CPU(s) list:             0-31
Vendor ID:                       AuthenticAMD
Model name:                      AMD EPYC 7313P 16-Core Processor
CPU family:                      25
Model:                           1
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        1
Frequency boost:                 enabled
CPU max MHz:                     3729.4919
CPU min MHz:                     1500.0000
BogoMIPS:                        5989.06
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Virtualization:                  AMD-V
L1d cache:                       512 KiB (16 instances)
L1i cache:                       512 KiB (16 instances)
L2 cache:                        8 MiB (16 instances)
L3 cache:                        128 MiB (4 instances)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.1.0.dev20230603
[pip3] torchaudio==2.1.0.dev20230603
[pip3] torchvision==0.16.0.dev20230603
[pip3] triton==2.1.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0           py310h7f8727e_0
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0
[conda] mkl_random                1.2.2           py310h00e6091_0
[conda] numpy                     1.24.3          py310hd5efca6_0
[conda] numpy-base                1.24.3          py310h8e6c178_0
[conda] pytorch                   2.1.0.dev20230603 py3.10_cuda11.8_cudnn8.7.0_0    pytorch-nightly
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchaudio                2.1.0.dev20230603     py310_cu118    pytorch-nightly
[conda] torchtriton               2.1.0+9820899b38           py310    pytorch-nightly
[conda] torchvision               0.16.0.dev20230603     py310_cu118    pytorch-nightly

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305

**Comment by ezyang** [comment]
torch.Size supports numel method and we apparently have not implemented it. Should be straightforward to add support for.

**Comment by soumya1729** [comment]
@KareemYousrii @ezyang this issue is fixed right? Seems to work ok on torch==2.0.1+cu118 with numel.


```
import torch

@torch.compile(fullgraph=True, mode='reduce-overhead')
def num():
    a = torch.randn(1, 2, 3, 4, 5)
    return torch.numel(a)

print("--no of elements: ",num())

```

**Comment by ezyang** [comment]
Your sample code is different; it looks like here you need to support `torch.Size([1, 2]).numel()` and similar.
